pkg/skaffold/kubernetes/debugging/adapter/adapter.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2021 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package adapter
+
+import (
+	v1 "k8s.io/api/core/v1"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/debug/types"
+)
+
+type Adapter struct {
+	container  *v1.Container
+	executable *types.ExecutableContainer
+}
+
+func NewAdapter(c *v1.Container) *Adapter {
+	return &Adapter{
+		container:  c,
+		executable: ExecutableContainerFromK8sContainer(c),
+	}
+}
+
+func (a *Adapter) GetContainer() *types.ExecutableContainer {
+	return a.executable
+}
+
+// Apply takes the relevant fields from the operable container
+// and applies them to the referenced v1.Container from the manifest's pod spec
+func (a *Adapter) Apply() {
+	a.container.Args = a.executable.Args
+	a.container.Command = a.executable.Command
+	a.container.Env = containerEnvToK8sEnv(a.executable.Env)
+	a.container.Ports = containerPortsToK8sPorts(a.executable.Ports)
+}
+
+// ExecutableContainerFromK8sContainer creates an instance of an operableContainer
+// from a v1.Container reference. This object will be passed around to accept
+// transforms, and will eventually overwrite fields from the creating v1.Container
+// in the manifest-under-transformation's pod spec.
+func ExecutableContainerFromK8sContainer(c *v1.Container) *types.ExecutableContainer {
+	return &types.ExecutableContainer{
+		Command: c.Command,
+		Args:    c.Args,
+		Env:     k8sEnvToContainerEnv(c.Env),
+		Ports:   k8sPortsToContainerPorts(c.Ports),
+	}
+}
+
+func k8sEnvToContainerEnv(k8sEnv []v1.EnvVar) types.ContainerEnv {
+	// TODO(nkubala): ValueFrom is ignored. Do we care?
```

Comment:
`ValueFrom` does matter: see #3872 and #4351

Reply:
ah ok. these will now be preserved during transformation